mlir/lib/Dialect/Rock/Tuning/RockTuningImpl.cpp

```diff
@@ -0,0 +1,111 @@
+    // XDLOPS
+    // M/block N/block K/block M/wave N/wave kPack aCopyMore bCopyMore
+    const std::vector<std::vector<uint32_t>> tParams = {{4, 8, 16, 32, 64, 128},
+                                                        {16, 32, 64, 128},
+                                                        {16, 32, 64, 128},
+                                                        {16, 32, 64},
+                                                        {16, 32, 64},
+                                                        {1, 4}};
+
+    for (uint32_t gemmMPerBlock : tParams[0]) {
+      for (uint32_t gemmNPerBlock : tParams[1]) {
+        for (uint32_t gemmKPerBlock : tParams[2]) {
+          for (uint32_t gemmMPerWave : tParams[3]) {
+            for (uint32_t gemmNPerWave : tParams[4]) {
+              for (uint32_t gemmKPack : tParams[5]) {
+                XdlopsGemmParamsAttr gemmParams =
+                    b.getAttr<XdlopsGemmParamsAttr>(
+                        gemmKPerBlock, gemmMPerBlock, gemmNPerBlock, gemmKPack,
+                        gemmMPerWave, gemmNPerWave);
+                newSpace->tuningRange.push_back(
+                    gemmParams.cast<RockTuningParamAttrInterface>());
+              }
+            }
+          }
+        }
+      }
+    }
+  } else {
+    // Non-XDLOPS
+    // M/block N/block K/block M/thread N/thread
+    const std::vector<std::vector<uint32_t>> tParams = {
+        {32, 64, 128}, {32, 64, 128}, {4, 8, 16}, {2, 4}, {2, 4}};
+    for (uint32_t gemmMPerBlock : tParams[0]) {
+      for (uint32_t gemmNPerBlock : tParams[1]) {
+        for (uint32_t gemmKPerBlock : tParams[2]) {
+          for (uint32_t gemmMPerThread : tParams[3]) {
+            for (uint32_t gemmNPerThread : tParams[4]) {
+              GeneralGemmParamsAttr gemmParams =
+                  b.getAttr<GeneralGemmParamsAttr>(
+                      gemmKPerBlock, gemmMPerBlock, gemmNPerBlock,
+                      /*kPerThread=*/1, gemmMPerThread, gemmNPerThread,
+                      /*kpack=*/1);
+              newSpace->tuningRange.push_back(
+                  gemmParams.cast<RockTuningParamAttrInterface>());
+            }
+          }
+        }
+      }
+    }
+  }
+
+  newSpace->numHeuristicQuick = 0;
+}
+
+TunableParams *createTunableParams(ModuleOp &mod) {
+  struct TunableParams *newSpace;
+  newSpace = new TunableParams();
+
+  // create range and heuristic
+  bool bFound = false;
+  mod->walk([&](rock::RockGemmWrapperInterface op) {
+    if (!bFound) {
+      bFound = true;
+      createGemmTuningRangeBF(newSpace, op);
+      newSpace->primaryOpType = op.getKernelType();
+    }
+  });
+
+  return newSpace;
+}
+
+bool tuningSetParam(ModuleOp &mod, ParamEntry *paramEntry) {
+  bool bFound = false;
+  mod->walk([&](rock::RockGemmWrapperInterface op) {
+    if (!bFound) {
+      bFound = true;
+      auto ctx = op.getContext();
+      StringAttr attr = StringAttr::get(ctx, paramEntry->perfString);
```

Comment:
Why not have the perf_config be a string attribute in the first place?

Reply:
Right, that sounds better.
setParam is likely to be called more often than getParamStr from the users.

Reply:
Uhm, just found there is a problem. A ParamEntry itself cannot have an mlir context to permanently stick with. So I can't create a StringAttr for a ParamEntry. ParamEntry will be created independently without tuning space (not yet available now).